pkg/operator/operator.go

```diff
@@ -44,13 +50,50 @@ func RunOperator(ctx context.Context, controllerConfig *controllercmd.Controller
 		return err
 	}
 
+	utilruntime.PanicHandlers = append(utilruntime.PanicHandlers, func(panicValue interface{}) {
+		degradedCond := operv1.OperatorCondition{
+			Type:    operv1.OperatorStatusTypeDegraded,
+			Status:  operv1.ConditionTrue,
+			Reason:  "ReconcileError",
+			Message: fmt.Sprintf("Panic detected: %v", panicValue),
+		}
+
+		// update operator status
+		_, _, err := v1helpers.UpdateStatus(ctx, o.client.Default().OperatorHelperClient(), v1helpers.UpdateConditionFn(degradedCond))
+		if err != nil {
+			klog.Errorf("Failed to update operator degraded condition on panic %v", err)
+			return
+		}
+
+		// update cluster operator status
+		err = retry.RetryOnConflict(retry.DefaultBackoff, func() error {
+			co := &configv1.ClusterOperator{}
+			if err := o.client.Default().CRClient().Get(context.TODO(), types.NamespacedName{Name: "network"}, co); err != nil {
+				return err
+			}
+			cohelpers.SetStatusCondition(&co.Status.Conditions, operstatus.OperatorConditionToClusterOperatorCondition(degradedCond))
+			if err := o.client.Default().CRClient().Status().Update(context.TODO(), co); err != nil {
+				return err
+			}
+			return nil
+		})
+		if err != nil {
+			klog.Errorf("Failed to update cluster operator degraded condition on panic %v", err)
+			return
+		}
+
+		panic(panicValue)
```

Comment:
We still want to panic instead of suppressing it?

Reply:
I think so. Consistency of the process itself cannot be ensured after a panic. The panic handling should do relatively simple things that if gone wrong don't risk the system so it's even bold that we attempt to update the status to degraded.
@kyrtapz can you point to a reference of other operators doing this for my peace of mind?

Reply:
I think we want to forward the panic as we are not really recovering from it here, we are just setting the status.
There is no assurance about the state of CNO after the crash occurs so  we should not allow it to continue, potentially causing more issues.

library-go controller wrapper(which CNO doesn't use) handles panics using `runtime.HandleCrash`: https://github.com/openshift/library-go/blob/a467dafcef6988afbcce43400c339d7ca68a3298/pkg/controller/factory/base_controller.go#L80
The panic will be forwarded if `runtime.ReallyCrash` is true: https://github.com/openshift/library-go/blob/db40400cfd41917bd76b1177023e3e5c69e9aa62/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go#L54
It's value can be [controlled](https://github.com/openshift/library-go/blob/9575de9b95b2542f69a00ff81ac4518b50cfdafe/pkg/serviceability/panic.go#L35) by an env variable: https://github.com/openshift/library-go/blob/4a2b231285b3a45bdbd3d86e2374965c626c3fae/pkg/controller/controllercmd/cmd.go#L115 (`false` by default)

CNO uses upstream controller runtime right now and none of its controllers implement panic recovery.
An alternative to the proposed approach would be to add panic handling to every controller manually.

Reply:
Wondering if we would lose info if we don't go through status manager `setDegraded`?
Also, there we do create or update and I am not sure if we need that here.

What would be the reasons not to go through status manager? Aside from redudinc the surface of further panics?

Reply:
Alright, thanks for extensive explanations! It makes sense to me.

Reply:
> What would be the reasons not to go through status manager? Aside from reducing the surface of further panics?

That was my original intent but after analyzing it further and talking to Jaime I think it is better to rely on status manager. In the previous approach there was a chance that while handling the panic (running deferred funcs) some other controller would clean the degraded status, using statusmanager with a new statusLevel addresses that issue.